File: src/app/lib/plc/runtime.ts

export class TON {
  t = 0;
  Q = false;

  update(dt: number, IN: boolean, PT: number) {
    if (IN) {
      this.t += dt;
      this.Q = this.t >= PT;
    } else {
      this.t = 0;
      this.Q = false;
    }
    return this.Q;
  }
}

export class TOF {
  t = 0;
  Q = true;

  update(dt: number, IN: boolean, PT: number) {
    if (IN) {
      this.Q = true;
      this.t = 0;
    } else {
      this.t += dt;
      if (this.t >= PT) this.Q = false;
    }
    return this.Q;
  }
}

export class TP {
  t = 0;
  Q = false;
  trig = false;

  update(dt: number, IN: boolean, PT: number) {
    if (IN && !this.trig) {
      this.trig = true;
      this.Q = true;
      this.t = 0;
    }

    if (this.trig) {
      this.t += dt;
      if (this.t >= PT) {
        this.Q = false;
        this.trig = false;
      }
    }

    return this.Q;
  }
}

export class PID {
  Kp = 1;
  Ki = 0;
  Kd = 0;

  i = 0;
  last = 0;

  update(dt: number, sp: number, pv: number) {

    const err = sp - pv;

    this.i += err * dt;

    const d = (err - this.last) / dt;

    const out =
      this.Kp * err +
      this.Ki * this.i +
      this.Kd * d;

    this.last = err;

    return out;
  }
}

export class PLC {
  env: any = {
    memory: {},
    fb: {},
    timers: {}
  };

  step(program: Function, dt: number) {
    program(this.env, dt);
  }
}

File: src/app/lib/plc/st2lua/runtime.lua

TON = {}
TON.__index = TON

function TON:new()
  return setmetatable({t=0,Q=false}, self)
end

function TON:update(dt, IN, PT)
  if IN then
    self.t = self.t + dt
    self.Q = self.t >= PT
  else
    self.t = 0
    self.Q = false
  end
  return self.Q
end


TOF = {}
TOF.__index = TOF

function TOF:new()
  return setmetatable({t=0,Q=true}, self)
end

function TOF:update(dt, IN, PT)
  if IN then
    self.t = 0
    self.Q = true
  else
    self.t = self.t + dt
    if self.t >= PT then
      self.Q = false
    end
  end
  return self.Q
end


TP = {}
TP.__index = TP

function TP:new()
  return setmetatable({t=0,Q=false,trig=false}, self)
end

function TP:update(dt, IN, PT)
  if IN and not self.trig then
    self.trig = true
    self.Q = true
    self.t = 0
  end

  if self.trig then
    self.t = self.t + dt
    if self.t >= PT then
      self.Q = false
      self.trig = false
    end
  end

  return self.Q
end
